### PR TITLE
Adds LRU Cache based Memoization, Weak Memoization, and Smtlib2 Parsing

### PR DESCRIPTION
**Pull request by tm507211**
This pull request implements two forms of weak memoization. One based on a resizable LRU cache, and another based on Weak Hashtbl (OCaml's Ephemerons).

It also updates the Smtlib2 support to include both outputting of SAT queries and parsing the response of such queries.

### CODE REVIEW

File: src/srkSmtlib2.ml [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+      | _ -> assert false
+    in
+    let reset _ =
+      int_id := -1;
+      real_id := -1;
+      bool_id := -1
+    in
+    { new_symbol; reset }
+
+  let rec simplify_const_expr expr =
```
Comment (zkincaid) [inline src/srkSmtlib2.ml]: This looks like it belongs in srkSimplify.ml.  Is there functionality in this simplfier that isn't present in the srkSimplify simplification routines / SimplifyingContext in syntax.ml?

File: src/srkSmtlib2Defs.ml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  | Real of QQ.t
+  | String of string
+
+type symbol = string
+
+type index =
+  | Num of ZZ.t
+  | Sym of symbol
+
+type identifier = symbol * (index list)
```
Comment (zkincaid) [inline src/srkSmtlib2Defs.ml:30]: This is a surprising type.  Can you explain in a comment?

File: src/syntax.mli [hunk 3]
```diff
@@ -345,8 +348,12 @@ val mk_false : 'a context -> 'a formula
    symbols or quantifiers. *)
 val eliminate_ite : 'a context -> 'a formula -> 'a formula
 
-(** Print a formula as a satisfiability query in SMTLIB2 format.  The query
-    includes function declarations and (check-sat). *)
+(** Print a formula as a satisfiability query in SMTLIB2 format.
+    The query includes function declarations and (check-sat). *)
+val pp_smtlib2_gen : ?named:bool -> ?env:(string Env.t) -> ?strings:((string, symbol) Hashtbl.t) ->
```
Comment (zkincaid) [inline src/syntax.mli:365]: What are the "named' and "strings" parameters?

File: src/srkSmtlib2.ml [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+        end
+      | srk_sym ->
+        begin match terms with
+        | [] ->
+          let sym = of_symbol srk_sym in
+          begin match typ_symbol srk sym with
+          | `TyBool
+          | `TyInt
+          | `TyReal -> (mk_const srk (of_symbol srk_sym) :> 'a gexpr)
+          | `TyFun _ -> invalid_arg "Unespected function symbol"
```
Comment (zkincaid) [inline src/srkSmtlib2.ml]: typo

File: src/srkSmtlib2Defs.ml (hunk 2 above)
Comment (tm507211) [reply src/srkSmtlib2Defs.ml:30]: This is the type of identifiers in Smtlib2. See http://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2017-07-18.pdf bottom of page 96. I'll add a comment in the code as well. An identifier is either a name (symbol) or a name followed by some indexes (e.g. "Vector Int 4" for a 4D vector of integers).

File: src/syntax.mli (hunk 3 above)
Comment (tm507211) [reply src/syntax.mli:365]: Named is whether or not each formula output should be named. E.g. the first assertion is named f0, then f1, ..., the nth one is named fn. If the formulas/assertions are named then you can use Smtlib2 commands to retrieve which assertions form the UNSAT core.

strings will store the mapping from Smtlib2 symbols/names to SRK symbols. This helps facilitate writing a Smtlib2 file, calling a SMT solver then retrieving the model and parsing it to a SRK interpretation.

File: src/srkSmtlib2.ml (hunk 1 above)
Comment (tm507211) [reply src/srkSmtlib2.ml]: You're right. This code does constant folding, which is covered by SimplifyingContext. Although, on careful examination, it seems the simplifying code in SimplifyingContext and this code differ on Mod.

Specifically, I think there's a typo in the SimplifyingContext code. It checks if the denominator is zero twice. I think the second check should actually be if the denominator is one and return zero. I can post a fix for that, but I have a question, do we support non-integral semantics for mod? e.g. can you do 1.2 mod 1 = 0.2?

**Pull request merged**